File: c/wasps.c
/* "Go to the Ant" - Wasp Task Differentiation, C port.
 *
 * Source: H. Van Dyke Parunak, "'Go to the Ant': Engineering Principles from
 * Natural Multi-Agent Systems," Annals of Operations Research 75:69-101 (1997),
 * discussing Theraulaz et al. 1991 (Polistes wasp task differentiation, §3.4).
 *
 * THE LENS (C): the colony state is a raw pair of arrays (F[], sig[]); the
 * mechanism laid utterly bare with manual memory and no abstraction between
 * agent and field. Every rule is index arithmetic over flat double arrays.
 *
 * Mature Polistes wasps -- genetically IDENTICAL -- split into a single Chief,
 * a band of Foragers, and a band of Nurses, with no HR department and no wasp
 * computing the proportion. Three interacting rules:
 *   1. FACE-OFFS. When two wasps meet, j beats i with the Fermi probability
 *      p = 1/(1 + e^(h*(F_i - F_j))). A quantum of Force passes loser -> winner.
 *   2. BROOD DEMAND.  D(t) = D(t-1) + appetite - W, W = food-work by foragers.
 *   3. FORAGE?  A wasp near the brood forages with p = 1/(1 + e^(hf*(sig_j - D))).
 *      Foraging LOWERS its threshold sig by xi (learning); not foraging RAISES
 *      it by phi (forgetting). Force is MOBILITY: a low-force wasp is stimulated
 *      by the brood but cannot travel to hunt.
 *
 * The joint (Force, Threshold) distribution self-separates into three castes:
 *   - Foragers  = high force, low threshold  (strong enough to move + sensitive)
 *   - Nurses    = low force,  low threshold  (attentive, but stuck near brood)
 *   - Chief     = one wasp, high force, high threshold (grounds the scales)
 *
 * PROVENANCE (carried from the Python reference):
 *   - The two Fermi formulas are PAPER VERBATIM (§3.4), tagged below.
 *   - The entropy-leak force bound (leak/gen, replacing an ad-hoc force cap) is
 *     OPERATIONALIZED (§4.6): a steady leak+gen bounds the hierarchy naturally
 *     (equilibrium mean ~ gen/leak). Honest tradeoff: dissipating force couples
 *     into the demand/threshold balance, so the Chief's high-threshold detail
 *     regresses vs a capped version.
 *   - dominance = (F/Fmax)^4 is OPERATIONALIZED as a spatiality proxy: the
 *     paper's Chief wanders and faces off (NOT near the brood), so it is rarely
 *     stimulated and its threshold drifts HIGH. (F/Fmax)^4 ~= 1 only for the
 *     single top wasp, leaving foragers (F ~ 0.7*Fmax) almost untouched, which
 *     restores the Chief's high-sigma caste.
 *
 * Dependency-free: C standard library only.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <math.h>

/* --- SplitMix64 PRNG (identical across all ports, seeded from --seed) ---
 * This is a CROSS-PORT CONVENTION; it intentionally does NOT reproduce CPython's
 * random module. All ports agree with EACH OTHER by consuming the RNG in the
 * same order, not with Python. */
static uint64_t rng_state;
static inline uint64_t rng_next(void) {
    rng_state += 0x9E3779B97F4A7C15ULL;
    uint64_t z = rng_state;
    z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9ULL;
    z = (z ^ (z >> 27)) * 0x94D049BB133111EBULL;
    return z ^ (z >> 31);
}
static inline double rng_float(void) { /* [0,1) */
    return (double)(rng_next() >> 11) * (1.0 / 9007199254740992.0); /* 2^-53 */
}
static inline uint64_t rng_randrange(uint64_t n) { return rng_next() % n; }
static inline double rng_uniform(double a, double b) {
    return a + rng_float() * (b - a);
}

/* --- Colony constants (Theraulaz §3.4 parameterisation) --- */
#define SIGMAX 4.0

/* --- The colony state: raw arrays, the mechanism laid bare. --- */
static double F[4096];    /* force (mobility) per wasp */
static double sig[4096];  /* foraging threshold per wasp */

/* comparator for qsort on doubles (ascending), used only for medians */
static int cmp_double(const void *a, const void *b) {
    double da = *(const double *)a, db = *(const double *)b;
    return (da > db) - (da < db);
}

int main(int argc, char **argv) {
    long ticks = 4000, n = 80;
    uint64_t seed = 0;
    for (int i = 1; i < argc; i++) {
        if (!strcmp(argv[i], "--seed") && i + 1 < argc) seed = strtoull(argv[++i], NULL, 10);
        else if (!strcmp(argv[i], "--ticks") && i + 1 < argc) ticks = strtol(argv[++i], NULL, 10);
        /* Python names this flag --wasps; we accept --ants (the shared CLI name) too. */
        else if ((!strcmp(argv[i], "--wasps") || !strcmp(argv[i], "--ants")) && i + 1 < argc)
            n = strtol(argv[++i], NULL, 10);
    }
    if (n < 1) n = 1;
    if (n > 4096) n = 4096;
    rng_state = seed;

    /* colony parameters (match the Python defaults) */
    const double h = 1.1, hf = 3.0, quantum = 0.10;
    const double appetite = 0.075 * (double)n;
    const double xi = 0.02, phi = 0.012, mob = 1.6;
    const double leak = 0.004, gen = 0.005;
    double D = 2.0;

    /* genetically identical: tiny initial spread only.
     * RNG order matches Python exactly: all n F-inits, then all n sig-inits. */
    for (long k = 0; k < n; k++) F[k] = 1.0 + rng_uniform(-0.05, 0.05);
    for (long k = 0; k < n; k++) sig[k] = 1.6 + rng_uniform(-0.05, 0.05);

    /* Forager/Nurse split history, sampled like the Python (every ticks//12). */
    long sample = ticks / 12; if (sample < 1) sample = 1;
    int hist_f[64], hist_n[64], nhist = 0;

    for (long t = 0; t < ticks; t++) {
        /* rule 1: face-offs -- gentle, capped, so a graded hierarchy forms. */
        for (long f = 0; f < n / 3; f++) {
            long i = (long)rng_randrange((uint64_t)n);
            long j = (long)rng_randrange((uint64_t)n);
            if (i == j) continue;
            /* PAPER §3.4 VERBATIM: p = 1/(1 + e^(h*(F_i - F_j))) */
            double pj = 1.0 / (1.0 + exp(h * (F[i] - F[j])));
            long w, l;
            if (rng_float() < pj) { w = j; l = i; } else { w = i; l = j; }
            double q = quantum < F[l] ? quantum : F[l];
            F[w] += q; F[l] -= q;                 /* force conserved in the face-off (paper) */
        }
        /* ENTROPY LEAK (§4.6, OPERATIONALIZED): force dissipates and regenerates.
         * A steady leak+gen bounds the hierarchy naturally (mean ~ gen/leak), so
         * no ad-hoc force cap is needed to stop one super-wasp. */
        for (long k = 0; k < n; k++) {
            double v = F[k] * (1.0 - leak) + gen;
            F[k] = v > 0.0 ? v : 0.0;
        }
        /* rules 2 & 3: brood stimulation + foraging. Work = COUNT of mobile foragers. */
        double Fmax = F[0];
        for (long k = 1; k < n; k++) if (F[k] > Fmax) Fmax = F[k];
        if (Fmax <= 0.0) Fmax = 1.0;
        long W = 0;
        for (long k = 0; k < n; k++) {
            /* PAPER §3.4 VERBATIM: p = 1/(1 + e^(hf*(sig - D))) */
            double pf = 1.0 / (1.0 + exp(hf * (sig[k] - D)));
            double ratio = F[k] / Fmax;
            /* SPATIALITY PROXY (OPERATIONALIZED): dom ~= 1 only for the single top wasp */
            double dom = ratio * ratio * ratio * ratio;
            if (rng_float() < pf * (1.0 - dom)) {  /* stimulated AND not away dominating */
                double s = sig[k] - xi;            /* learns: threshold drops */
                sig[k] = s > 0.0 ? s : 0.0;
                if (F[k] > mob) W++;               /* mobile enough to actually hunt */
            } else {
                double s = sig[k] + phi;           /* forgets: threshold rises */
                sig[k] = s < SIGMAX ? s : SIGMAX;
            }
        }
        D = D + appetite - (double)W;
        if (D < 0.0) D = 0.0;

        if (t % sample == 0 && nhist < 64) {
            /* count Forager/Nurse under the current caste rule */
            double sorted_sig[4096];
            memcpy(sorted_sig, sig, sizeof(double) * (size_t)n);
            qsort(sorted_sig, (size_t)n, sizeof(double), cmp_double);
            double smed = sorted_sig[n / 2];
            long chief = 0;
            for (long k = 1; k < n; k++) if (F[k] > F[chief]) chief = k;
            int nf = 0, nn = 0;
            for (long k = 0; k < n; k++) {
                if (k == chief) continue;
                if (F[k] > mob && sig[k] <= smed) nf++;
                else nn++;
            }
            hist_f[nhist] = nf; hist_n[nhist] = nn; nhist++;
        }
    }

    /* --- Final castes --- */
    long chief = 0;
    for (long k = 1; k < n; k++) if (F[k] > F[chief]) chief = k;
    double sorted_sig[4096];
    memcpy(sorted_sig, sig, sizeof(double) * (size_t)n);
    qsort(sorted_sig, (size_t)n, sizeof(double), cmp_double);
    double smed = sorted_sig[n / 2];

    /* per-caste tallies: Chief (the argmax wasp), Forager, Nurse */
    int cnt[3] = {0, 0, 0};
    double sumF[3] = {0, 0, 0}, sumS[3] = {0, 0, 0};
    double popF = 0.0;
    for (long k = 0; k < n; k++) {
        popF += F[k];
        int g;
        if (k == chief) g = 0;
        else if (F[k] > mob && sig[k] <= smed) g = 1;  /* Forager: mobile + responsive */
        else g = 2;                                     /* Nurse: immobile or unresponsive */
        cnt[g]++; sumF[g] += F[k]; sumS[g] += sig[k];
    }

    printf("Emergent castes from %ld genetically identical wasps (%ld ticks):\n\n", n, ticks);
    const char *names[3] = {"Chief", "Forager", "Nurse"};
    for (int g = 0; g < 3; g++) {
        if (cnt[g] == 0) continue;
        printf("  %-8s n=%3d   mean Force %5.2f   mean Threshold %5.2f\n",
               names[g], cnt[g], sumF[g] / cnt[g], sumS[g] / cnt[g]);
    }
    printf("\n  Chief force %.2f (pop mean %.2f), threshold %.2f\n",
           F[chief], popF / (double)n, sig[chief]);
    printf("  Forager/Nurse split(t):");
    for (int i = 0; i < nhist; i++) printf(" %d/%d", hist_f[i], hist_n[i]);
    printf("\n");

    /* --- (F,sigma) ASCII landscape: x = Force ->, y = Threshold ^ --- */
    const int cols = 48, rows = 16;
    double fmn = F[0], fmx = F[0], smn = sig[0], smx = sig[0];
    for (long k = 1; k < n; k++) {
        if (F[k] < fmn) fmn = F[k];
        if (F[k] > fmx) fmx = F[k];
        if (sig[k] < smn) smn = sig[k];
        if (sig[k] > smx) smx = sig[k];
    }
    double sorted_F[4096];
    memcpy(sorted_F, F, sizeof(double) * (size_t)n);
    qsort(sorted_F, (size_t)n, sizeof(double), cmp_double);
    double fmed = sorted_F[n / 2];
    char grid[16][48];
    memset(grid, ' ', sizeof grid);
    for (long k = 0; k < n; k++) {
        int x = (int)((F[k] - fmn) / (fmx - fmn + 1e-9) * (cols - 1));
        int y = (int)((sig[k] - smn) / (smx - smn + 1e-9) * (rows - 1));
        if (x < 0) x = 0;
        if (x > cols - 1) x = cols - 1;
        if (y < 0) y = 0;
        if (y > rows - 1) y = rows - 1;
        char mark = (k == chief) ? 'C' : (F[k] >= fmed ? 'F' : 'n');
        grid[rows - 1 - y][x] = mark;
    }
    printf("\n  (F,sigma) landscape - x = Force ->, y = Threshold ^ | C chief, F forager, n nurse:\n\n");
    for (int r = 0; r < rows; r++) {
        printf("   ");
        for (int col = 0; col < cols; col++) putchar(grid[r][col]);
        putchar('\n');
    }
    return 0;
}

File: wasps.py
#!/usr/bin/env python3
r"""'Go to the Ant' §3.4 — Wasp Task Differentiation (Theraulaz et al. 1991), recreated from the paper.

Mature Polistes wasps — genetically IDENTICAL — split into a single Chief, a band of Foragers, and a band
of Nurses, with no HR department and no wasp computing the proportion. Parunak's three interacting rules:

  1. FACE-OFFS. When two wasps meet, j beats i with the Fermi probability  p = 1/(1 + e^(h·(F_i − F_j))).
     The higher force usually wins (but not always); a quantum of Force passes loser → winner.
  2. BROOD DEMAND.  D(t) = D(t−1) + appetite − W, where W is the food-work done by all foragers.
  3. FORAGE?  A wasp near the brood forages with Fermi probability  p = 1/(1 + e^(h·(σ_j − D))).
     Foraging LOWERS its threshold σ by ξ (learning); not foraging RAISES σ by φ (forgetting).

Force is MOBILITY (a low-force wasp is stimulated by the brood but cannot travel to hunt). The joint
(Force, Threshold) distribution self-separates into three clusters — the castes:
  · Foragers  = high force, low threshold  (strong enough to move + sensitive to the brood)
  · Nurses    = low force,  low threshold  (attentive, but stuck near the brood)
  · Chief     = one wasp, high force, high threshold (grounds the scales via face-offs; doesn't forage)

  python wasps.py                 # caste populations + the (F,σ) landscape as it settles
"""
import argparse, math, random


class Colony:
    SIGMAX = 4.0
    # NOTE (provenance, honest labelling): the genuine §4.6 entropy leak for wasps is RULE 1's conservative
    # force TRANSFER among wasps (Parunak names "the flow of force among wasps that drives the emergence of the
    # three roles"). The leak/gen term below is a SEPARATE Force-RELAXATION (mean-reversion to ~gen/leak) — it
    # replaces an ad-hoc force CAP and bounds the hierarchy, but it is an INFERENCE BEYOND Parunak (plausibly a
    # Theraulaz 1991 element we can't verify; the primary source isn't on hand), NOT the §4.6 entropy leak. It is
    # empirically required: pure conservation condenses to one super-wasp with no graded forager band. See PROVENANCE.md.
    def __init__(self, n=80, seed=0, h=1.1, hf=3.0, quantum=0.10, appetite=None,
                 xi=0.02, phi=0.012, mob=1.6, leak=0.004, gen=0.005):
        self.rng = random.Random(seed); self.n = n
        # genetically identical: tiny initial spread only
        self.F = [1.0 + self.rng.uniform(-0.05, 0.05) for _ in range(n)]
        self.sig = [1.6 + self.rng.uniform(-0.05, 0.05) for _ in range(n)]
        self.seenmax = list(self.F)   # each wasp's LOCAL, FADING memory of the top force it has faced
        self.seendecay = 0.998        # the memory fades, so a stale early peak can't stick (robustness)
        self.D = 2.0
        self.h, self.hf, self.q = h, hf, quantum
        # appetite sized so the mobile minority (~n/8) can ALMOST meet demand — leaving the top wasp
        # (the Chief) surplus, so its threshold drifts high while the working foragers stay low.
        self.appetite = appetite if appetite is not None else 0.075 * n
        self.xi, self.phi, self.mob = xi, phi, mob
        self.leak, self.gen = leak, gen

    def step(self):
        rng, F, sig, n = self.rng, self.F, self.sig, self.n
        # rule 1: face-offs — gentle, capped, so a graded hierarchy forms (not one super-wasp)
        for _ in range(n // 3):
            i, j = rng.randrange(n), rng.randrange(n)
            if i == j:
                continue
            fi, fj = F[i], F[j]
            pj = 1.0 / (1.0 + math.exp(self.h * (fi - fj)))      # PAPER §3.4 VERBATIM: p=1/(1+e^(h(Fi-Fj)))
            w, l = (j, i) if rng.random() < pj else (i, j)
            t = min(self.q, F[l])
            F[w] += t; F[l] -= t                         # force is conserved in the face-off (paper)
            m = fi if fi > fj else fj                    # LOCAL: each wasp's FADING memory of the strongest force it
            di = self.seenmax[i] * self.seendecay        # has faced — a gossip estimate of the top force, NO global
            dj = self.seenmax[j] * self.seendecay        # max. Fading lets the emergent Chief's own force become its
            self.seenmax[i] = m if m > di else di        # remembered peak (dom->1), so its threshold reliably drifts
            self.seenmax[j] = m if m > dj else dj        # high (0/32 seed failures vs 3/32 for the global-max version)
        # FORCE RELAXATION (inference beyond Parunak — NOT the §4.6 entropy leak; that is Rule 1's force flow
        # above): force mean-reverts toward ~gen/leak each tick. A steady leak+gen bounds the hierarchy naturally
        # (equilibrium mean ≈ gen/leak), so no ad-hoc force cap is needed to stop one super-wasp.
        for k in range(n):
            F[k] = max(0.0, F[k] * (1.0 - self.leak) + self.gen)
        # rules 2 & 3: brood stimulation + foraging. Work = COUNT of mobile foragers (each brings 1).
        # SPATIALITY PROXY (operationalized, now fully LOCAL): the paper's Chief "wanders and faces off", so it
        # is NOT near the brood and is rarely stimulated -> its threshold drifts HIGH. We approximate "away
        # dominating" by suppressing a wasp's foraging in proportion to how it ranks against the strongest force
        # IT HAS FACED (self.seenmax, a per-agent gossip estimate — NO global max): dominance=(F/seenmax)^4 hits
        # ~1 only for the wasp that out-forces everyone it meets (the Chief), leaving foragers almost untouched.
        W = 0
        for k in range(n):
            pf = 1.0 / (1.0 + math.exp(self.hf * (sig[k] - self.D)))  # PAPER §3.4 VERBATIM: p=1/(1+e^(h(sig-D)))
            dom = (F[k] / (self.seenmax[k] or 1.0)) ** 4          # ~1 only for the wasp on top of its own encounters
            if rng.random() < pf * (1.0 - dom):                   # stimulated AND not away dominating
                sig[k] = max(0.0, sig[k] - self.xi)               # learns: threshold drops
                if F[k] > self.mob:                               # mobile enough to actually hunt
                    W += 1
            else:
                sig[k] = min(self.SIGMAX, sig[k] + self.phi)      # forgets: threshold rises
        self.D = max(0.0, self.D + self.appetite - W)

    def castes(self):
        F, sig = self.F, self.sig
        chief = max(range(self.n), key=lambda k: F[k])
        smed = sorted(sig)[self.n // 2]
        groups = {"Chief": [], "Forager": [], "Nurse": []}
        for k in range(self.n):
            if k == chief:
                groups["Chief"].append(k)
            elif F[k] > self.mob and sig[k] <= smed:
                groups["Forager"].append(k)      # mobile + responsive
            else:
                groups["Nurse"].append(k)         # immobile (or unresponsive) -> stays with the brood
        return groups, chief


def run(ticks=4000, n=80, seed=0, verbose=True):
    c = Colony(n=n, seed=seed)
    hist = []
    for t in range(ticks):
        c.step()
        if t % max(1, ticks // 12) == 0:
            g, _ = c.castes()
            hist.append((len(g["Forager"]), len(g["Nurse"])))
    if verbose:
        g, chief = c.castes()
        print(f"Emergent castes from {n} genetically identical wasps ({ticks} ticks):\n")
        for name in ("Chief", "Forager", "Nurse"):
            ks = g[name]
            if not ks:
                continue
            mF = sum(c.F[k] for k in ks) / len(ks); mS = sum(c.sig[k] for k in ks) / len(ks)
            print(f"  {name:8} n={len(ks):3d}   mean Force {mF:5.2f}   mean Threshold {mS:5.2f}")
        print(f"\n  Chief force {c.F[chief]:.2f} (pop mean {sum(c.F)/n:.2f}), "
              f"threshold {c.sig[chief]:.2f}")
        print("  Forager/Nurse split(t):", " ".join(f"{f}/{ns}" for f, ns in hist))
        landscape(c)
    return c


def landscape(c, cols=48, rows=16):
    """ASCII scatter of the population in (Force -> x, Threshold -> y) space."""
    F, sig = c.F, c.sig
    fmn, fmx = min(F), max(F); smn, smx = min(sig), max(sig)
    grid = [[" "] * cols for _ in range(rows)]
    chief = max(range(c.n), key=lambda k: F[k])
    fmed = sorted(F)[c.n // 2]
    for k in range(c.n):
        x = int((F[k] - fmn) / (fmx - fmn + 1e-9) * (cols - 1))
        y = int((sig[k] - smn) / (smx - smn + 1e-9) * (rows - 1))
        mark = "C" if k == chief else ("F" if F[k] >= fmed else "n")
        grid[rows - 1 - y][x] = mark
    print("\n  (F,σ) landscape — x = Force →, y = Threshold ↑ | C chief, F forager, n nurse:\n")
    for row in grid:
        print("   " + "".join(row))


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--ticks", type=int, default=4000)
    ap.add_argument("--wasps", type=int, default=80)
    ap.add_argument("--seed", type=int, default=0)
    a = ap.parse_args()
    run(a.ticks, a.wasps, a.seed)


if __name__ == "__main__":
    main()
